File: ts-mcp-client/src/sampling.ts
import Anthropic from '@anthropic-ai/sdk';

import { DEEPSEEK_API_KEY, DEEPSEEK_BASE_URL, DEEPSEEK_MODEL, HAS_KEY } from './config.js';

type ContentBlock = { type: string; text?: string; [k: string]: unknown };

export interface SampleParams {
  messages: { role: 'user' | 'assistant'; content: ContentBlock | ContentBlock[] }[];
  maxTokens?: number;
  systemPrompt?: string;
}

export interface SampleResult {
  role: 'assistant';
  content: { type: 'text'; text: string };
  model: string;
  stopReason: string;
}

function contentToText(content: ContentBlock | ContentBlock[] | undefined): string {
  const blocks = Array.isArray(content) ? content : content ? [content] : [];
  return blocks
    .map((b) =>
      b && b.type === 'text' && typeof b.text === 'string'
        ? b.text
        : `[${b?.type ?? 'unknown'} content]`,
    )
    .join('\n');
}

/** DeepSeek via its Anthropic-compatible endpoint (the real path when DEEPSEEK_API_KEY is set). */
async function sampleWithDeepSeek(params: SampleParams): Promise<SampleResult> {
  const client = new Anthropic({ apiKey: DEEPSEEK_API_KEY, baseURL: DEEPSEEK_BASE_URL });
  const resp = await client.messages.create({
    model: DEEPSEEK_MODEL,
    max_tokens: params.maxTokens ?? 512,
    system: params.systemPrompt,
    messages: params.messages.map((m) => ({ role: m.role, content: contentToText(m.content) })),
  });
  const text = resp.content
    .filter((b): b is Anthropic.TextBlock => b.type === 'text')
    .map((b) => b.text)
    .join('');
  return {
    role: 'assistant',
    content: { type: 'text', text },
    model: resp.model,
    stopReason: resp.stop_reason ?? 'endTurn',
  };
}

/** Deterministic stand-in so Sampling works before a key is configured. */
function sampleMock(params: SampleParams): SampleResult {
  const lastUser = [...params.messages].reverse().find((m) => m.role === 'user');
  const said = contentToText(lastUser?.content).replace(/\s+/g, ' ').trim();
  const gist = said.split(' ').slice(0, 16).join(' ');
  return {
    role: 'assistant',
    content: {
      type: 'text',
      text: `(mock model — set DEEPSEEK_API_KEY in backend/.env for a real DeepSeek answer)\nIn short: ${gist}${said.split(' ').length > 16 ? '…' : ''}`,
    },
    model: 'mock-deepseek',
    stopReason: 'endTurn',
  };
}

export async function sample(params: SampleParams): Promise<SampleResult> {
  return HAS_KEY ? sampleWithDeepSeek(params) : sampleMock(params);
}

File: py-mcp-client/sampling.py
"""Sampling — the client runs a model on the server's behalf (``sampling/createMessage``).

Routes to DeepSeek via its Anthropic-compatible endpoint when ``DEEPSEEK_API_KEY`` is set;
otherwise a deterministic mock so Sampling works before a key is configured.
"""

from __future__ import annotations

import httpx

from config import DEEPSEEK_API_KEY, DEEPSEEK_BASE_URL, DEEPSEEK_MODEL, HAS_KEY


def _content_to_text(content: object) -> str:
  blocks = content if isinstance(content, list) else [content] if content else []
  out = []
  for b in blocks:
    if isinstance(b, dict) and b.get("type") == "text" and isinstance(b.get("text"), str):
      out.append(b["text"])
    else:
      out.append(f"[{b.get('type', 'unknown') if isinstance(b, dict) else 'unknown'} content]")
  return "\n".join(out)


def _sample_with_deepseek(params: dict) -> dict:
  """DeepSeek via its Anthropic-compatible endpoint (the real path when a key is set)."""
  messages = [
    {"role": m.get("role", "user"), "content": _content_to_text(m.get("content"))}
    for m in params.get("messages", [])
  ]
  body = {"model": DEEPSEEK_MODEL, "max_tokens": params.get("maxTokens") or 512, "messages": messages}
  if params.get("systemPrompt"):
    body["system"] = params["systemPrompt"]
  resp = httpx.post(
    f"{DEEPSEEK_BASE_URL}/v1/messages",
    headers={"x-api-key": DEEPSEEK_API_KEY, "anthropic-version": "2023-06-01", "content-type": "application/json"},
    json=body,
    timeout=60.0,
  ).raise_for_status().json()
  text = "".join(b.get("text", "") for b in resp.get("content", []) if b.get("type") == "text")
  return {
    "role": "assistant",
    "content": {"type": "text", "text": text},
    "model": resp.get("model", DEEPSEEK_MODEL),
    "stopReason": resp.get("stop_reason") or "endTurn",
  }


def _sample_mock(params: dict) -> dict:
  """Deterministic stand-in so Sampling works before a key is configured."""
  last_user = next((m for m in reversed(params.get("messages", [])) if m.get("role") == "user"), None)
  said = " ".join(_content_to_text(last_user.get("content") if last_user else None).split())
  words = said.split(" ")
  gist = " ".join(words[:16])
  ellipsis = "…" if len(words) > 16 else ""
  return {
    "role": "assistant",
    "content": {
      "type": "text",
      "text": f"(mock model — set DEEPSEEK_API_KEY in py-mcp-client/.env for a real DeepSeek answer)\nIn short: {gist}{ellipsis}",
    },
    "model": "mock-deepseek",
    "stopReason": "endTurn",
  }


def sample(params: dict) -> dict:
  """Run sampling against DeepSeek (when keyed) or the deterministic mock."""
  if HAS_KEY:
    try:
      return _sample_with_deepseek(params)
    except Exception:  # noqa: BLE001 — fall back to the mock on any provider error
      return _sample_mock(params)
  return _sample_mock(params)
